mongo/client_bulk_write.go

```diff
@@ -0,0 +1,703 @@
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.Acknowledged {
+		if mb.result.DeleteResults == nil {
+			mb.result.DeleteResults = make(map[int]ClientBulkWriteDeleteResult)
+		}
+		mb.result.DeleteResults[idx] = ClientBulkWriteDeleteResult{int64(cur.N)}
+	}
+
+	return true
+}
+
+func (mb *modelBatches) appendInsertResult(cur *cursorInfo, raw bson.Raw) bool {
+	idx := int(cur.Idx) + mb.offset
+	if err := cur.extractError(); err != nil {
+		err.Raw = raw
+		if mb.writeErrors == nil {
+			mb.writeErrors = make(map[int]WriteError)
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.Acknowledged {
+		if mb.result.InsertResults == nil {
+			mb.result.InsertResults = make(map[int]ClientBulkWriteInsertResult)
+		}
+		mb.result.InsertResults[idx] = ClientBulkWriteInsertResult{mb.newIDMap[idx]}
+	}
+
+	return true
+}
+
+func (mb *modelBatches) appendUpdateResult(cur *cursorInfo, raw bson.Raw) bool {
+	idx := int(cur.Idx) + mb.offset
+	if err := cur.extractError(); err != nil {
+		err.Raw = raw
+		if mb.writeErrors == nil {
+			mb.writeErrors = make(map[int]WriteError)
+		}
+		mb.writeErrors[idx] = *err
+		return false
+	}
+
+	if mb.result.Acknowledged {
+		if mb.result.UpdateResults == nil {
+			mb.result.UpdateResults = make(map[int]ClientBulkWriteUpdateResult)
+		}
+		result := ClientBulkWriteUpdateResult{
+			MatchedCount: int64(cur.N),
+		}
+		if cur.NModified != nil {
+			result.ModifiedCount = int64(*cur.NModified)
+		}
+		if cur.Upserted != nil {
+			result.UpsertedID = cur.Upserted.ID
+		}
+		mb.result.UpdateResults[idx] = result
+	}
+
+	return true
+}
+
+type clientInsertDoc struct {
+	namespace int
+	document  interface{}
+}
+
+func (d *clientInsertDoc) marshal(bsonOpts *options.BSONOptions, registry *bson.Registry) (interface{}, bsoncore.Document, error) {
+	uidx, doc := bsoncore.AppendDocumentStart(nil)
+
+	doc = bsoncore.AppendInt32Element(doc, "insert", int32(d.namespace))
+	f, err := marshal(d.document, bsonOpts, registry)
+	if err != nil {
+		return nil, nil, err
+	}
+	var id interface{}
+	f, id, err = ensureID(f, bson.NilObjectID, bsonOpts, registry)
+	if err != nil {
+		return nil, nil, err
+	}
+	doc = bsoncore.AppendDocumentElement(doc, "document", f)
+	doc, err = bsoncore.AppendDocumentEnd(doc, uidx)
+	return id, doc, err
+}
+
+type clientUpdateDoc struct {
+	namespace      int
+	filter         interface{}
+	update         interface{}
+	hint           interface{}
+	arrayFilters   []interface{}
+	collation      *options.Collation
+	upsert         *bool
+	multi          bool
+	checkDollarKey bool
+}
+
+func (d *clientUpdateDoc) marshal(bsonOpts *options.BSONOptions, registry *bson.Registry) (bsoncore.Document, error) {
+	uidx, doc := bsoncore.AppendDocumentStart(nil)
+
+	doc = bsoncore.AppendInt32Element(doc, "update", int32(d.namespace))
+
+	f, err := marshal(d.filter, bsonOpts, registry)
+	if err != nil {
+		if d.filter == nil {
+			return nil, fmt.Errorf("%w: filter is required", err)
+		}
```

Comment:
Is there a reason to nest this nil filter check in this error handling block? Can we move this check to the beginning of `marshal`?

Optional: Consider a clearer error message, like
> update filter cannot be nil